main/lsp/requests/go_to_implementation.cc

```diff
@@ -0,0 +1,113 @@
+    if (!isAbstract) {
+        return make_unique<ResponseError>(
+            (int)LSPErrorCodes::InvalidParams,
+            "Go to implementation can be used only for methods or references of abstract classes");
+    }
+
+    auto childClasses = findChildren(gs, owningClassSymbolRef);
+    auto methodName = method.data(gs)->name;
+    for (const auto &childClass : childClasses) {
+        auto methodImplementation = childClass.data(gs)->findMember(gs, methodName);
+        locations.push_back(methodImplementation.data(gs)->loc());
+    }
+
+    return locations;
+}
+
+GoToImplementationTask::GoToImplementationTask(const LSPConfiguration &config, MessageId id,
+                                               std::unique_ptr<ImplementationParams> params)
+    : LSPRequestTask(config, move(id), LSPMethod::TextDocumentImplementation), params(move(params)) {}
+
+unique_ptr<ResponseMessage> GoToImplementationTask::runRequest(LSPTypecheckerDelegate &typechecker) {
+    auto response = make_unique<ResponseMessage>("2.0", id, LSPMethod::TextDocumentImplementation);
+
+    const core::GlobalState &gs = typechecker.state();
+    auto queryResult =
+        queryByLoc(typechecker, params->textDocument->uri, *params->position, LSPMethod::TextDocumentImplementation);
+
+    if (queryResult.error) {
+        // An error happened while setting up the query.
+        response->error = move(queryResult.error);
+        return response;
+    }
+
+    if (queryResult.responses.empty()) {
+        return response;
+    }
+
+    vector<unique_ptr<Location>> result;
+    auto queryResponse = move(queryResult.responses[0]);
+    if (auto def = queryResponse->isDefinition()) {
+        // User called "Go to Implementation" from the abstract function definition
+        core::SymbolRef method = def->symbol;
+        auto locationsOrError = findMethodImplementations(gs, method);
+
+        if (auto error = get_if<unique_ptr<ResponseError>>(&locationsOrError)) {
+            response->error = move(*error);
+            return response;
+        } else if (auto locations = get_if<vector<core::Loc>>(&locationsOrError)) {
+            for (const auto &location : *locations) {
+                addLocIfExists(gs, result, location);
+            }
+        }
+    } else if (auto constant = queryResponse->isConstant()) {
+        // User called "Go to Implementation" from the abstract class reference
+        auto classSymbol = constant->symbol;
+
+        if (!classSymbol.data(gs)->isClassOrModuleAbstract()) {
+            response->error = make_unique<ResponseError>(
+                (int)LSPErrorCodes::InvalidParams,
+                "Go to implementation can be used only for methods or references of abstract classes");
+            return response;
+        }
+
+        auto childClasses = findChildren(gs, classSymbol);
+        for (const auto &childClass : childClasses) {
+            addLocIfExists(gs, result, childClass.data(gs)->loc());
```

Comment:
Classes can be reopened in multiple files, so they may have multiple locs. You may want to have an inner loop here:
```suggestion
            for (auto loc : childClass.data(gs)->locs()) {
              addLocIfExists(gs, result, loc);
            }
```

Reply:
Surprisingly, this didn't work out. `Symbol::locs` store only two values. As I understood, it should be the primary location of the symbol and optional rbi file. I've added new property, which stores all the possible locations of the module/class.

However, I'm not sure how hard it will hit the memory consumptions. Would it be a large issue, how do you think?